test/decl/protocol/req/associated_type_generic_param.swift

```diff
@@ -1,10 +1,11 @@
 // RUN: %target-typecheck-verify-swift -enable-experimental-associated-type-inference
-// RUN: not %target-typecheck-verify-swift -disable-experimental-associated-type-inference
+// RUN: %target-typecheck-verify-swift -disable-experimental-associated-type-inference
 
 protocol P {
   associatedtype A = Int
 }
 
 struct S<A>: P {}
 
-let x: String.Type = S<String>.A.self
+// This is unfortunate but it is the behavior of Swift 5.10.
+let x: Int.Type = S<String>.A.self
```

Comment:
Would looking for generic parameters before defaults but after same-type requirements not fix this at least?

Reply:
Maybe, but I would rather fix this in one shot by building a disjunction containing the default associated type and generic parameter, if any, in inferTypeWitnessesViaAssociatedType(), and having the solver walk through the solution space. This first requires better ranking rules and a smarter solver that won't go exponential as easily.